tardis/montecarlo/src/integrator.c

```diff
@@ -0,0 +1,259 @@
+}
+indexpair_t nu_idx_from_nu_pair(double nu_blu, double nu_red, const double* line_nu, int len)
+{
+    indexpair_t pair = { .start = 0, .end = len-1};
+    bool found_blu = false;
+
+    for (int idx = 0; idx < len; ++idx)
+    {
+        if ((line_nu[idx] <= nu_blu) & (!found_blu)){
+            pair.start = idx;
+            found_blu = true;}
+        if (line_nu[idx] <= nu_red){
+            pair.end = idx;
+            break;}
+    }
+
+    return pair;
+}
+
+double intensity_black_body(double nu, double T){
+    double k_B_cgs = 1.3806488000e-16;
+    double h_cgs   = 6.6260695700e-27;
+    double c_cgs   = 2.9979245800e+10;
+    double coefficient, beta_rad;
+    beta_rad = 1 / (k_B_cgs * T);
+    coefficient = 2 * h_cgs / (c_cgs * c_cgs);
+    return coefficient * nu*nu*nu / (exp(h_cgs * nu * beta_rad) -1 );
+}    
+
+double get_r(int cr_idx, int no_of_cr_shells, const double* Rs)
+{
+    int r_idx;
+    if (cr_idx < no_of_cr_shells){
+        r_idx = cr_idx;}
+    else if (cr_idx < 2*no_of_cr_shells){
+        r_idx = no_of_cr_shells-1 - cr_idx % no_of_cr_shells;}
+
+    return Rs[r_idx];
+}
+
+int get_cr_sign(int cr_idx, int no_of_cr_shells)
+{
+    if (cr_idx < no_of_cr_shells){
+        return -1;}
+    else if (cr_idx < 2*no_of_cr_shells){
+        return 1;}
+}
+
+int get_cr_start(int no_of_shells, double p, double R_ph)
+{
+    if (p >= R_ph) {
+        return 0;}
+    else if (p < R_ph) {
+        return no_of_shells+1;
+    }
+}
+
+int get_sh_idx(int cr_idx, int no_of_cr_shells)
+{
+    if (cr_idx < no_of_cr_shells){
+       return cr_idx;}
+    else if (cr_idx < 2*no_of_cr_shells-1){
+        return (2*(no_of_cr_shells-1) - cr_idx);}
+
+}
+int get_num_shell_cr(double p, const double* Rs, int len)
+{
+    assert(Rs[0] > Rs[1]);
+
+    int num;
+    for(num = 0; num < len; ++num)
+    {
+        if( p >= Rs[num]) {
+            break;}
+    }
+    return num;
+}
+double sum_lines(indexpair_t nu_lims, double I_nu, const double* taus, const double* att_S_ul, int sh_idx, int len)
+{
+    double result = I_nu;
+    for (int k_idx = nu_lims.start; k_idx <= nu_lims.end; ++k_idx)
+    {
+        result = result * exp(-taus[k_idx*len +sh_idx]) + GET_IJ(att_S_ul,sh_idx,k_idx,len);
+    }
+    return result;
+}
+double integrate_intensity(const double* I_nu, const double* ps, int len)
+{
+    double result = 0.0;
+    double h = (ps[0]-ps[len-1])/len;
+    result =  I_nu[0]*h*ps[0]/2.0;
+
+    for (int idx = 1; idx < len-1; ++idx)
+    {
+        result += I_nu[idx]*h*ps[idx];
+    }
+    result += I_nu[len-1]*h*ps[len-1]/2.0;
+    return result*8*M_PI*M_PI;
+}
+void integrate_source_functions(double* L_nu, const double* line_nu, const double* taus, const double* att_S_ul, const double* nus,
+        const double* ps, const double* Rs, double T, double R_ph, double inv_ct, const int64_t* lens)
+{
+    double* I_nu  = calloc(lens[PLEN], sizeof(double));
+    int no_of_cr_shells, sh_idx, cr_start, cr_end;
+    int kludge = 0;
```

Comment:
I forgot, but it also shouldn't matter since it serves to fix a off-by-one problem with the indexing once our p-rays cross the photosphere. Since we never leave the photosphere after entering it during correct execution, resetting it probably isn't an issue.

But yeah, a kludge to get the shell indexing working correctly in the photosphere. The python version doesn't count the photosphere as a shell that is crossed, but it is natural to do so in the c version since it is one loop: however, because of this mismatch I had to put in this fix to get the C version to index the shells correctly since I based much on the python version.